### PR TITLE
debezium/dbz#2379 Add pluggable claim-check storage

### PR DESCRIPTION
**Pull request by omkarmehta06**
Fixes debezium/dbz#2379

## Description

This change adds an opt-in `claim_check` strategy to the `EnforceRecordSize` SMT while preserving proportional truncation as the default behavior.

- Adds a storage-neutral `OversizedRecordStorage` SPI in `debezium-storage-api`.
- Adds `S3OversizedRecordStorage` in `debezium-storage-s3`, using the AWS default credentials provider chain.
- Serializes and synchronously stores the complete source record before emitting a versioned marker.
- Uses deterministic source-position and payload hashes so retries reuse the same object key.
- Fails the transformation without mutating the input record when storage fails or the replacement still exceeds `max.bytes`.
- Documents configuration, marker recovery, S3 setup, and fail-closed behavior.

## Verification

- Focused `EnforceRecordSize` suites: 29 tests passed.
- Focused S3 adapter suite: 7 tests passed.
- Affected reactor `verify`: all 20 modules passed, including 397 connect-plugin tests, 34 S3 tests, formatting, checkstyle, architecture checks, and API compatibility.

## AI assistance disclosure

I used OpenAI Codex assistively to review contribution hygiene, run tests, and resolve a documentation conflict while rebasing onto current `main`. I reviewed the resulting changes and take responsibility for this contribution.

## PR Checklist

- [ ] I have read the [contribution guidelines](https://github.com/debezium/debezium/blob/main/CONTRIBUTING.md) and the [governance document](https://github.com/debezium/governance/blob/main/GOVERNANCE.md) on PR expectations.
- [ ] Minimal changes to code not directly related to this change.
- [ ] One feature/change per PR unless tightly coupled.
- [ ] Rebased on upstream `main`.

### CODE REVIEW

File: debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+import software.amazon.awssdk.services.s3.S3ClientBuilder;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+
+/**
+ * Stores complete oversized records in Amazon S3.
+ *
+ * @author Debezium Authors
+ */
+@Incubating
+public class S3OversizedRecordStorage implements OversizedRecordStorage {
```
Comment (Naros) [inline debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java:35]: For consistency with other S3 components like the schema history and the post processor, we should consider adding `access.key.id` and `secret.access.key` support here. WDYT?

File: debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    private String prefix;
+    private S3Client s3Client;
+
+    @Override
+    public void configure(Configuration config) {
+        close();
+
+        URI basePath = parseBasePath(config.getString(BASE_PATH_CONFIG));
+        String regionName = config.getString(REGION_CONFIG);
+        if (regionName == null || regionName.isBlank()) {
```
Comment (Naros) [inline debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java]:
```suggestion
        if (Strings.isNullOrBlank(regionName)) {
```
There are several places where the `Strings` utility class should be preferred rather than the blank, empty, or null checks. Could you review the full PR and apply `io.debezium.util.Strings` consistently?

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java [hunk 3]
```diff
@@ -149,7 +243,153 @@ public R apply(R record) {
...
+        org.apache.kafka.connect.data.Field sectionField = envelope.schema().field(sectionName);
+        if (sectionField == null || sectionField.schema().type() != Schema.Type.STRUCT) {
+            return;
+        }
+        for (org.apache.kafka.connect.data.Field field : sectionField.schema().fields()) {
+            if (claimCheckColumns.contains(field.name())
+                    && field.schema().type() != Schema.Type.STRING
+                    && field.schema().type() != Schema.Type.BYTES) {
+                throw new ConnectException("Claim-check column '" + field.name() + "' has schema type "
+                        + field.schema().type() + "; only STRING and BYTES columns are supported");
```
Comment (Naros) [inline debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java]: If we take a Postgres table with a column `amount NUMERIC(38,2)`, using the `precise` decimal mode, Debezium emits this as an `org.apache.kafka.connect.data.Decimal` that uses `BYTES`. If a user lists this amounts column in the `claim.check.columns.include.list`, this method accepts it because this only checks that the type is not BYTES.

However, when `claimCheckMaker` returns the JSON as `byte[]`, the call to `section.put` will throw an exception that its an invalid object.

Now it may be unlikely that someone would put this specific type in the claim check, but there are potentially other use cases where this could report an unusual error to the user, particularly given that this class' `unqualifiedColumnName` stripping removes any database and/or schema reference, and two tables that share the same column name but with very different types could easily manifest this issue.

The fix I'd recommend is small: treat a BYTES field with a non-null schema name as unsupported.

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/ClaimCheckRecordSerializer.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    static SerializedRecord serialize(SourceRecord record) {
+        Map<String, Object> payload = new LinkedHashMap<>();
+        payload.put("version", 1);
+        payload.put("topic", record.topic());
+        payload.put("sourcePartition", canonicalMap(record.sourcePartition()));
+        payload.put("sourceOffset", canonicalMap(record.sourceOffset()));
+        payload.put("key", connectValue(record.keySchema(), record.key()));
+        payload.put("value", connectValue(record.valueSchema(), record.value()));
```
Comment (Naros) [inline debezium-connect-plugins/src/main/java/io/debezium/transforms/ClaimCheckRecordSerializer.java:47]: Should `headers` also be serialized?

**Review by Naros — COMMENTED** [review]
hi @omkarmehta06 great PR, I've left a few inline comments.

File: debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java (hunk 1 above)
Comment (omkarmehta06) [reply debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java:35]: Thanks @Naros. Our production deployment uses IRSA, so we need the default AWS credentials provider chain to remain the normal path. We can support both modes: use `access.key.id` and `secret.access.key` when both are configured, and otherwise retain `DefaultCredentialsProvider`, including IRSA. The optional static creds will make local development against s3-compatible endpoints easier. Let me support and cover both paths.

File: debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java (hunk 2 above)
Comment (omkarmehta06) [reply debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java]: I'm handling this via commit. Initially, I applied this suggestion from UI and got an error: `Please prefix the commit message(s) with the debezium/dbz#xxx`
I'll be applying `io.debezium.util.Strings` in the full PR. It makes sense. This is much cleaner.

File: debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java (hunk 1 above)
Comment (omkarmehta06) [reply debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java:35]: Addressed in the next commit ([1725564d2](https://github.com/debezium/debezium/commit/1725564d2a6bbda065863fdc68303d1ee4703672)): optional access.key.id and secret.access.key support is added, with fallback to the default AWS provider chain when they are absent or incomplete.

File: debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java (hunk 2 above)
Comment (omkarmehta06) [reply debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java]: Done

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/ClaimCheckRecordSerializer.java (hunk 4 above)
Comment (omkarmehta06) [reply debezium-connect-plugins/src/main/java/io/debezium/transforms/ClaimCheckRecordSerializer.java:47]: Headers are now serialized as an ordered list, preserving duplicate names and their order.

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java (hunk 3 above)
Comment (omkarmehta06) [reply debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java]: I really appreciate you raising this and the thorough review. Addressed in the next commit ([1725564d2](https://github.com/debezium/debezium/commit/1725564d2a6bbda065863fdc68303d1ee4703672)): claim-check now rejects named BYTES logical schemas, including Kafka Connect Decimal, before any storage write. This makes the supported contract explicit: claim-check markers support strings and raw, unnamed binary values, not logical values encoded as bytes.

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java [hunk 5]
```diff
@@ -149,7 +244,156 @@ public R apply(R record) {
...
+            if (claimCheckColumns.contains(field.name()) && !isSupportedClaimCheckSchema(field.schema())) {
+                throw new ConnectException("Claim-check column '" + field.name() + "' has schema type "
+                        + field.schema().type() + "; only STRING and unnamed BYTES columns are supported");
+            }
+        }
+    }
+
+    private static boolean isSupportedClaimCheckSchema(Schema schema) {
+        return schema.type() == Schema.Type.STRING
+                || (schema.type() == Schema.Type.BYTES && schema.name() == null);
```
Comment (Naros) [inline debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java]: Following on from my earlier `Decimal` comment — thanks for fixing that — the BYTES branch is now guarded by `schema.name() == null`, but the STRING branch is not, and I think the STRING case is actually the worse of the two.

Debezium emits several **named STRING** logical types: `io.debezium.data.Json`, `Xml`, `Uuid`, `EnumSet`, and `io.debezium.time.ZonedTimestamp` (all `SchemaBuilder.string().name(...)` — see `SchemaFactory:358-361` and `ZonedTimestamp:103-104`).

If a user lists such a column, `isSupportedClaimCheckSchema` accepts it. Unlike the `Decimal` case, `Struct.put` will not object, because `ConnectSchema.validateValue` only checks that the value is a `String`. The marker JSON is written straight into, say, a `ZonedTimestamp` column and a downstream consumer reads `{"__debezium_claim_check":true,...}` where a timestamp should be — with no error anywhere in the pipeline.

If the intent is "raw values only, not logical types", the guard would be symmetric:

```java
return schema.name() == null
        && (schema.type() == Schema.Type.STRING || schema.type() == Schema.Type.BYTES);
```

I have deliberately not made that a one-click suggestion, because it is a behaviour change and you may have had a reason to allow named strings — if so, the error message at line 352 and the documented contract should say so explicitly. Either way, a test for a named STRING column alongside `shouldRejectNamedBytesColumnsBeforeWriting` would lock the decision in.

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java [hunk 6]
```diff
@@ -149,7 +244,156 @@ public R apply(R record) {
...
+        collectSectionColumns(envelope, "after", availableColumns);
+
+        Set<String> missingColumns = claimCheckColumns.stream()
+                .filter(column -> !availableColumns.contains(column))
+                .collect(Collectors.toCollection(LinkedHashSet::new));
+        if (!missingColumns.isEmpty()) {
+            throw new ConnectException("Claim-check columns " + missingColumns + " were not found in record for topic " + topic);
+        }
+        if (fields.isEmpty()) {
+            throw new ConnectException("Claim-check columns are present but null in record for topic " + topic);
```
Comment (Naros) [inline debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java]: An SMT is applied connector-wide, so these two throws are stricter than they may look.

With several captured tables, the first oversized event from a table that does not have the configured column fails the task — even though the SMT would otherwise have passed that record through untouched. The same happens when the configured columns are all present but null.

Fail-closed is exactly right when storage might silently drop data, but both of these fire *before* any storage interaction, on a record the transformation does not really apply to. Falling back to `truncate`, or passing through with a warning, when no configured column matches would be safer for the multi-table case.

If failing is deliberate, could the "Applies to" or limitations section spell out that `claim_check` requires every captured table to carry the configured columns? That constraint is not obvious from the current documentation.

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java [hunk 7]
```diff
@@ -282,11 +531,77 @@ public void configure(Map<String, ?> props) {
...
+            oversizedRecordStorage.configure(Configuration.from(storageProperties));
+            this.claimCheckStorage = oversizedRecordStorage;
+        }
+        catch (RuntimeException e) {
+            oversizedRecordStorage.close();
+            throw new ConfigException("Unable to configure claim-check storage class " + storageClass, e);
+        }
+    }
+
+    private static String unqualifiedColumnName(String columnName) {
```
Comment (Naros) [inline debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java]: This keeps only the segment after the last dot, so `inventory.customers.payload` becomes `payload` and then matches a `payload` column in any table the connector captures.

The documentation does state this, so it is not undocumented, but the option is named `claim.check.columns.include.list`, mirroring Debezium's `column.include.list`, where a fully-qualified name is scoped to a table. I think most users will supply a qualified name expecting scoping and instead silently get cross-table offloading.

`S3LargeMessagePostProcessor` uses `io.debezium.function.Predicates` for the equivalent matching, which would give qualified-regex semantics consistent with the rest of Debezium and would also make the two features behave alike.

WDYT?

File: debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        String endpoint = config.getString(ENDPOINT_CONFIG);
+        if (!Strings.isNullOrBlank(endpoint)) {
+            builder.endpointOverride(parseAbsoluteUri(ENDPOINT_CONFIG, endpoint));
+        }
+        return builder.build();
+    }
+
+    private static AwsCredentialsProvider createCredentialsProvider(Configuration config) {
+        String accessKeyId = config.getString(ACCESS_KEY_ID_CONFIG);
+        String secretAccessKey = config.getString(SECRET_ACCESS_KEY_CONFIG);
```
Comment (Naros) [inline debezium-storage/debezium-storage-s3/src/main/java/io/debezium/storage/s3/S3OversizedRecordStorage.java]: Now that static credentials are supported, the secret should carry `PASSWORD` typing the way the other S3 components do. Compare `S3SchemaHistory.java:74`:

```java
public static final Field SECRET_ACCESS_KEY = Field.create(CONFIGURATION_FIELD_PREFIX_STRING + SECRET_ACCESS_KEY_CONFIG)
        .withDisplayName("S3 secret access key")
        .withType(Type.PASSWORD)
        ...
```

Here the value is read straight off `Configuration` with no `Field` declaration at all. More broadly, none of the eight `claim.check.storage.*` properties appear in the SMT's `CONFIG_DEF` or in `getConfigFields()` — they are read directly from the raw `props` map, which works, but means they get no `PASSWORD` masking wherever Debezium or Connect keys off ConfigDef type, and no validation.

It also leaves `getConfigFields()` disagreeing with the options table you documented: the table lists all eight, while the programmatic metadata lists only `strategy`, `claim.check.storage.class`, and `claim.check.columns.include.list`.

Declaring the storage properties as `Field`s (with `Type.PASSWORD` for the secret) and adding them to `getConfigFields()` would line this up with `S3SchemaHistory` and with the docs.

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/ClaimCheckRecordSerializer.java [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    static String sha256Hex(byte[] data) {
+        try {
+            byte[] digest = MessageDigest.getInstance("SHA-256").digest(data);
+            StringBuilder hex = new StringBuilder(digest.length * 2);
+            for (byte value : digest) {
+                hex.append(String.format("%02x", value));
+            }
+            return hex.toString();
```
Comment (Naros) [inline debezium-connect-plugins/src/main/java/io/debezium/transforms/ClaimCheckRecordSerializer.java]: This loop could be replaced with `io.debezium.util.HexConverter.convertToHexString(byte[])`, as the utility will perform faster than `String.format`.

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java (hunk 6 above)
Comment (Naros) [inline debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java]: Could `claim_check` fall back to proportional truncation here rather than throwing?

The schema-based lookup is right — `collectSectionColumns` reads `sectionField.schema().fields()`, so this correctly does not misfire on deletes or on merely-unset columns; it only trips when the record's schema genuinely lacks the column, meaning a different table. Since an SMT is connector-wide, that is any oversized record from a table that does not carry the configured columns. Line 338 is the same situation with the columns present but null.

The inconsistency is with the other strategy in this same class: when truncation has nothing to act on it returns the record unchanged (line 230), whereas `claim_check` fails the task.

Falling back to `truncate` would make the two consistent, and it is better than either alternative I can see — the record still lands under `max.bytes`, so there is no task failure, and no `RecordTooLargeException` relocated to the producer either, which is what simply passing the record through would give you.

One caveat on relying on `errors.tolerance=all` instead: that only helps on Kafka Connect, where SMTs run through `RetryWithToleranceOperator` (`TransformationChain:58`). Debezium's embedded engine calls `apply` directly (`Transformations:92`), so Debezium Server has no per-record escape hatch.

No fail-closed concern either way — both checks run before any storage write, so nothing can be left half-stored.

**Review by Naros — COMMENTED** [review]
Hi @omkarmehta06 one final pass, and I found a few outliers. After this, I believe this can be merged.

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java [hunk 10]
```diff
@@ -6,9 +6,14 @@
 package io.debezium.transforms;
```
Comment (debezium-builder) [inline debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java:6]: Please put these two classes into a separate Java sub-package.

File: debezium-connect-plugins/src/test/java/io/debezium/transforms/EnforceRecordSizeClaimCheckTest.java [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.debezium.config.Configuration;
+import io.debezium.spi.storage.OversizedRecord;
+import io.debezium.spi.storage.OversizedRecordReference;
+import io.debezium.spi.storage.OversizedRecordStorage;
+
+class EnforceRecordSizeClaimCheckTest {
```
Comment (debezium-builder) [inline debezium-connect-plugins/src/test/java/io/debezium/transforms/EnforceRecordSizeClaimCheckTest.java:38]: Could we add a test for null key?

**Review by debezium-builder — COMMENTED** [review]
@omkarmehta06 LGTM, thanks! I left two small proposals.

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java (hunk 7 above)
Comment (omkarmehta06) [reply debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java]: Good catch. Locally, I will replace the final-segment lookup with a case-insensitive Predicates selector. Each pattern can be tested against both payload and inventory.customers.payload; a topic-qualified pattern should reject an oversized record from another topic before any storage write occurs. The docs and regression coverage will be included in the next commit.

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/ClaimCheckRecordSerializer.java (hunk 9 above)
Comment (omkarmehta06) [reply debezium-connect-plugins/src/main/java/io/debezium/transforms/ClaimCheckRecordSerializer.java]: I'll need to get familiar with all util.*
I will replace the per-byte `String.format` loop with `HexConverter.convertToHexString`

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java (hunk 10 above)
Comment (omkarmehta06) [reply debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java:6]: Could you clarify which two classes you intend here? The comment anchors at the existing `EnforceRecordSize` package, while the new claim-check types span the SMT helper and storage SPI. I want to move the intended types without changing the existing SMT public name or unnecessarily broadening the SPI move.

File: debezium-connect-plugins/src/test/java/io/debezium/transforms/EnforceRecordSizeClaimCheckTest.java (hunk 11 above)
Comment (omkarmehta06) [reply debezium-connect-plugins/src/test/java/io/debezium/transforms/EnforceRecordSizeClaimCheckTest.java:38]: Yes

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java (hunk 7 above)
Comment (omkarmehta06) [reply debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java]: Addressed in [541687fe3](https://github.com/debezium/debezium/commit/541687fe30fad1187550d20785a0b818e962d1f2). Thank you

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/ClaimCheckRecordSerializer.java (hunk 9 above)
Comment (omkarmehta06) [reply debezium-connect-plugins/src/main/java/io/debezium/transforms/ClaimCheckRecordSerializer.java]: Addressed in [541687fe3](https://github.com/debezium/debezium/commit/541687fe30fad1187550d20785a0b818e962d1f2). Thank you

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java (hunk 10 above)
Comment (omkarmehta06) [reply debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java:6]: Addressed in [541687fe3](https://github.com/debezium/debezium/commit/541687fe30fad1187550d20785a0b818e962d1f2). Thank you

Comment (jpechane) [reply debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java:6]: @omkarmehta06 At least ClaimCheckRecordSerializer should go into a separate package. We can think about moving `EnforceRecordSize` later.

**Review by jpechane — COMMENTED** [review]
@omkarmehta06 LGTM, thanks. I still belive that at least the serializer class should go into a separate package. When don this can be merged.

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java (hunk 10 above)
Comment (omkarmehta06) [reply debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java:6]: Addressed in [7883e09b8](https://github.com/debezium/debezium/commit/7883e09b8d7e8886650ff43e331c0f4eb38e7b07). Moved ClaimCheckRecordSerializer into io.debezium.transforms.claimcheck, keeping EnforceRecordSize in its existing package. Thanks for clarifying, Jiri.

Comment (omkarmehta06) [reply debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java:6]: Hi @jpechane, following up now that ClaimCheckRecordSerializer is in its own package in 7883e09b8. Could you take another look and let us know whether anything else is needed before merging? Thanks again for the review!

Comment (omkarmehta06) [reply debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java:6]: Also, CI failed while downloading the Unicode scanner with “Connection reset by peer,” which skipped the downstream tests. Could you please rerun the failed jobs?

File: debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java [hunk 12]
```diff
@@ -149,7 +250,172 @@ public R apply(R record) {
...
+            Predicate<String> selector = claimCheckColumnSelectors.get(index);
+            boolean matched = availableColumns.stream()
+                    .anyMatch(availableColumn -> matchesClaimCheckColumn(selector, topic, availableColumn));
+            if (!matched) {
+                missingColumns.add(claimCheckColumns.get(index));
+            }
+        }
+        if (!missingColumns.isEmpty()) {
+            throw new ConnectException("Claim-check columns " + missingColumns + " were not found in record for topic " + topic);
+        }
```
Comment (Naros) [inline debezium-connect-plugins/src/main/java/io/debezium/transforms/EnforceRecordSize.java:348]: I think this validation makes topic-qualified patterns unusable for more than one table. The loop requires every configured pattern to match a column in the current record and throws if any don't. With a configuration that targets two tables, such as

```
claim.check.columns.include.list=inventory[.]customers[.]payload,inventory[.]orders[.]notes
```

an oversized record from `inventory.customers` fails because the `orders` pattern has nothing to match in that record. I tried this and got:

```
ConnectException: Claim-check columns [inventory[.]orders[.]notes] were not found in record for topic inventory.customers
```

So after the scoping change, a qualified pattern effectively only works when a single table is configured. 

Related to that, `shouldScopeQualifiedClaimCheckColumnsToTheConfiguredTopic` asserts that an oversized record from a topic that no pattern targets throws as well. With the default `truncate` strategy that record would simply be truncated, whereas with `claim_check` one large row in a table that is not listed stops the connector.

Would it make sense to treat a pattern that does not apply to the current record as not applicable rather than as an error? When no pattern applies to a record, the SMT could either pass the record through or fall back to truncation.

**Review by Naros — COMMENTED** [review]
Hi @omkarmehta06 I found one small problem below. Could you take a look as you have time, otherwise this LGTM.
